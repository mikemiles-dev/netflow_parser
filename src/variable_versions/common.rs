use crate::protocol::ProtocolTypes;

use nom::bytes::complete::take;
use nom::error::{Error as NomError, ErrorKind};
use nom::number::complete::{be_u128, be_u32};
use nom::Err as NomErr;
use nom::IResult;
use nom_derive::*;
use serde::Serialize;

use std::net::{Ipv4Addr, Ipv6Addr};
use std::time::Duration;

/// Holds our datatypes and values post parsing
#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Serialize)]
#[serde(untagged)]
pub enum DataNumber {
    U8(u8),
    U16(u16),
    U32(u32),
    U64(u64),
    U128(u128),
    I32(i32),
}

/// Convert into usize, mainly for serialization purposes
impl DataNumber {
    pub fn get_value(self) -> usize {
        let result: usize = self.into();
        result
    }

    /// Parse bytes into DataNumber Type
    pub fn parse(i: &[u8], field_length: u16, signed: bool) -> IResult<&[u8], DataNumber> {
        match field_length {
            1 if !signed => Ok(u8::parse(i)?).map(|(i, j)| (i, Self::U8(j))),
            2 if !signed => Ok(u16::parse(i)?).map(|(i, j)| (i, Self::U16(j))),
            4 if signed => Ok(i32::parse(i)?).map(|(i, j)| (i, Self::I32(j))),
            4 if !signed => Ok(u32::parse(i)?).map(|(i, j)| (i, Self::U32(j))),
            8 if !signed => Ok(u64::parse(i)?).map(|(i, j)| (i, Self::U64(j))),
            16 if !signed => Ok(u128::parse(i)?).map(|(i, j)| (i, Self::U128(j))),
            _ => Err(NomErr::Error(NomError::new(i, ErrorKind::Fail))),
        }
    }

    pub fn from_field_type(
        remaining: &[u8],
        field_type: FieldDataType,
        field_length: u16,
    ) -> IResult<&[u8], FieldValue> {
        let (remaining, field_value) = match field_type {
            FieldDataType::UnsignedDataNumber => {
                let (i, data_number) = DataNumber::parse(remaining, field_length, false)?;
                (i, FieldValue::DataNumber(data_number))
            }
            FieldDataType::SignedDataNumber => {
                let (i, data_number) = DataNumber::parse(remaining, field_length, true)?;
                (i, FieldValue::DataNumber(data_number))
            }
            FieldDataType::String => {
                let (i, taken) = take(field_length)(remaining)?;
                (
                    i,
                    FieldValue::String(String::from_utf8_lossy(taken).to_string()),
                )
            }
            FieldDataType::Ip4Addr => {
                let (i, taken) = be_u32(remaining)?;
                let ip_addr = Ipv4Addr::from(taken);
                (i, FieldValue::Ip4Addr(ip_addr))
            }
            FieldDataType::Ip6Addr => {
                let (i, taken) = be_u128(remaining)?;
                let ip_addr = Ipv6Addr::from(taken);
                (i, FieldValue::Ip6Addr(ip_addr))
            }
            FieldDataType::DurationSeconds => {
                let (i, data_number) = DataNumber::parse(remaining, field_length, false)?;
                (
                    i,
                    FieldValue::Duration(Duration::from_secs(data_number.get_value() as u64)),
                )
            }
            FieldDataType::DurationMillis => {
                let (i, data_number) = DataNumber::parse(remaining, field_length, false)?;
                (
                    i,
                    FieldValue::Duration(Duration::from_millis(data_number.get_value() as u64)),
                )
            }
            FieldDataType::DurationMicros => {
                let (i, data_number) = DataNumber::parse(remaining, field_length, false)?;
                (
                    i,
                    FieldValue::Duration(Duration::from_micros(data_number.get_value() as u64)),
                )
            }
            FieldDataType::DurationNanos => {
                let (i, data_number) = DataNumber::parse(remaining, field_length, false)?;
                (
                    i,
                    FieldValue::Duration(Duration::from_nanos(data_number.get_value() as u64)),
                )
            }
            FieldDataType::ProtocolType => {
                let (i, protocol) = ProtocolTypes::parse(remaining)?;
                (i, FieldValue::ProtocolType(protocol))
            }
            FieldDataType::Float64 => {
                let (i, f) = f64::parse(remaining)?;
                (i, FieldValue::Float64(f))
            }
            FieldDataType::Vec => {
                let (i, taken) = take(field_length)(remaining)?;
                (i, FieldValue::Vec(taken.to_vec()))
            }
            FieldDataType::Unknown => {
                let (i, taken) = take(field_length)(remaining)?;
                (i, FieldValue::Vec(taken.to_vec()))
            }
        };
        Ok((remaining, field_value))
    }
}
/// Convert into usize, mainly for serialization purposes
impl From<DataNumber> for usize {
    fn from(val: DataNumber) -> Self {
        match val {
            DataNumber::U8(i) => i as usize,
            DataNumber::U32(i) => i as usize,
            DataNumber::I32(i) => i as usize,
            DataNumber::U16(i) => i as usize,
            DataNumber::U64(i) => i as usize,
            DataNumber::U128(i) => i as usize,
        }
    }
}

/// Holds the post parsed field with its relvant datatype
#[derive(Debug, PartialEq, PartialOrd, Clone, Serialize)]
pub enum FieldValue {
    String(String),
    DataNumber(DataNumber),
    Float64(f64),
    Duration(Duration),
    Ip4Addr(Ipv4Addr),
    Ip6Addr(Ipv6Addr),
    Vec(Vec<u8>),
    ProtocolType(ProtocolTypes),
    Unknown,
}

/// Helps the parser indefiy the data type to parse the field as
#[derive(Debug, PartialEq, Eq, Clone, Serialize)]
pub enum FieldDataType {
    String,
    SignedDataNumber,
    UnsignedDataNumber,
    Float64,
    DurationSeconds,
    DurationMillis,
    DurationMicros,
    DurationNanos,
    Ip4Addr,
    Ip6Addr,
    Vec,
    ProtocolType,
    Unknown,
}
