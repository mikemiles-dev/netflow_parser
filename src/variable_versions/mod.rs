pub mod data_number;
pub mod ipfix;
pub mod ipfix_lookup;
pub mod v9;
pub mod v9_lookup;
