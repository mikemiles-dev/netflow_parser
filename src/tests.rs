#[cfg(test)]
mod base_tests {

    use crate::variable_versions::ipfix::{
        Template as IPFixTemplate, TemplateField as IPFixTemplateField,
    };
    use crate::variable_versions::v9::{
        Template as V9Template, TemplateField as V9TemplateField,
    };
    use crate::NetflowParser;

    use hex;
    use insta::assert_yaml_snapshot;

    #[test]
    fn it_parses_unix_timestamp_correctly() {
        use nom::number::complete::{be_u32, be_u64};
        use std::time::Duration;

        let packet = [5, 0, 6, 7, 8, 9, 0, 1];
        let (remain, secs1) =
            be_u32::<&[u8], nom::error::Error<&[u8]>>(packet.as_slice()).unwrap();
        let (remain, nsecs1) = be_u32::<&[u8], nom::error::Error<&[u8]>>(remain).unwrap();
        assert_eq!(remain, []);

        let time1 = Duration::from_nanos(nsecs1 as u64) + Duration::from_secs(secs1 as u64);

        let (remain, secs_nsecs) =
            be_u64::<&[u8], nom::error::Error<&[u8]>>(packet.as_slice()).unwrap();
        assert_eq!(remain, []);
        let secs2 = (secs_nsecs >> 32) as u32 as u64;
        let nsecs2 = secs_nsecs as u32;

        let time2 = Duration::new(secs2, nsecs2);

        assert_eq!(secs1 as u64, secs2);
        assert_eq!(nsecs1, nsecs2);
        assert_eq!(time1, time2);
    }

    #[test]
    #[cfg(not(feature = "unix_timestamp"))]
    fn it_parses_v5() {
        let packet = [
            0, 5, 0, 1, 3, 0, 4, 0, 5, 0, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1, 2, 3,
            4, 5, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1,
            2, 3, 4, 5, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7,
        ];
        assert_yaml_snapshot!(NetflowParser::default().parse_bytes(&packet));
    }

    #[test]
    #[cfg(feature = "unix_timestamp")]
    fn it_parses_v5_timestamp() {
        let packet = [
            0, 5, 2, 0, 3, 0, 4, 0, 5, 0, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1, 2, 3,
            4, 5, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1,
            2, 3, 4, 5, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7,
        ];
        assert_yaml_snapshot!(NetflowParser::default().parse_bytes(&packet));
    }

    #[test]
    fn it_creates_error() {
        let packet = [12, 13, 14];
        assert_yaml_snapshot!(NetflowParser::default().parse_bytes(&packet));
    }

    #[test]
    #[cfg(not(feature = "unix_timestamp"))]
    fn it_parses_v7() {
        let packet = [
            0, 7, 0, 1, 3, 0, 4, 0, 5, 0, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1, 2, 3,
            4, 5, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1,
            2, 3, 4, 5, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1,
        ];
        assert_yaml_snapshot!(NetflowParser::default().parse_bytes(&packet));
    }

    #[test]
    #[cfg(feature = "unix_timestamp")]
    fn it_parses_v7_timestamp() {
        let packet = [
            0, 7, 2, 0, 3, 0, 4, 0, 5, 0, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1, 2, 3,
            4, 5, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1,
            2, 3, 4, 5, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1,
        ];
        assert_yaml_snapshot!(NetflowParser::default().parse_bytes(&packet));
    }

    #[test]
    fn it_parses_v9() {
        let packet = [
            0, 9, 0, 2, 0, 0, 9, 9, 0, 1, 2, 3, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 16, 1, 2, 0,
            2, 0, 1, 0, 4, 0, 8, 0, 4, 1, 2, 0, 12, 9, 2, 3, 4, 9, 9, 9, 8,
        ];
        assert_yaml_snapshot!(NetflowParser::default().parse_bytes(&packet));
    }

    #[test]
    fn it_parses_v9_no_data() {
        let packet = [
            0, 9, 0, 2, 0, 0, 9, 9, 0, 1, 2, 3, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 16, 1, 2, 0,
            2, 0, 1, 0, 4, 0, 8, 0, 4, 1, 2, 0, 4,
        ];
        assert_yaml_snapshot!(NetflowParser::default().parse_bytes(&packet));
    }

    #[test]
    fn it_parses_v9_template_and_data_packet() {
        let hex = r#"000900090000cc99664728d80000000100000000000000480400001000080004000c000400160004001500040001000400020004000a0004000e0004003d00010088000100070002000b00020004000100060001003c000100050001000000400401000e00080004000c000400160004001500040001000400020004000a0004000e0004003d0001008800010020000200040001003c0001000500010000004808000010001b0010001c001000160004001500040001000400020004000a0004000e0004003d00010088000100070002000b00020004000100060001003c000100050001000000400801000e001b0010001c001000160004001500040001000400020004000a0004000e0004003d000100880001008b000200040001003c0001000500010001001a01000004000c000200040022000400230001005200100100001d000000000000000101616e7900000000000000000000000000080000cd200300d1ef3b2200fe3497fffeb7686e26064700303200000000000068155338000016050000a418000004bf0000000900000000000000000003e2d401bb0619060026064700303200000000000068155338200300d1ef3b2200fe3497fffeb7686e000016050000a418000000b4000000030000000000000000010301bbe2d406040600000000000000000000000000000000010000000000000000000000000000000100004f2c00004f2c0000003c00000001000000000000000000031f91aaea06140600000000040001009df0fb3dc0a801d700006308000076d500000b7200000018000000000000000000031466ac50061a0400c0a801d79df0fb3d00006308000076d500000e370000002100000000000000000103ac501466061e04005dd10e499df0fb3d00006308000076d6000004bd0000000b00000000000000000003ac501466061e04009df0fb3d5dd10e4900006308000076d6000005b90000000c000000000000000001031466ac50061a0400c0a80125c6fcce190000a9180000a9cd0000027b0000000900000000000000000003a34401bb06190400c6fcce19c0a801250000a9180000a9cd00000108000000060000000000000000010301bba34406150400"#;
        let packet = hex::decode(hex).unwrap();
        assert_yaml_snapshot!(NetflowParser::default().parse_bytes(&packet));
    }

    #[test]
    fn it_parses_multiple_packets() {
        let v9_packet = [
            0, 9, 0, 2, 0, 0, 9, 9, 0, 1, 2, 3, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 16, 1, 2, 0,
            2, 0, 1, 0, 4, 0, 8, 0, 4, 1, 2, 0, 12, 9, 2, 3, 4, 9, 9, 9, 8,
        ];
        let v7_packet = [
            0, 7, 0, 1, 3, 0, 4, 0, 5, 0, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1, 2, 3,
            4, 5, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1,
            2, 3, 4, 5, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1,
        ];
        let v5_packet = [
            0, 5, 0, 1, 3, 0, 4, 0, 5, 0, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1, 2, 3,
            4, 5, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1,
            2, 3, 4, 5, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7,
        ];
        let ipfix_packet = [
            0, 10, 0, 64, 1, 2, 3, 4, 0, 0, 0, 0, 1, 2, 3, 4, 0, 2, 0, 20, 1, 0, 0, 3, 0, 8, 0,
            4, 0, 12, 0, 4, 0, 2, 0, 4, 1, 0, 0, 28, 1, 2, 3, 4, 1, 2, 3, 3, 1, 2, 3, 2, 0, 2,
            0, 2, 0, 1, 2, 3, 4, 5, 6, 7,
        ];
        let mut all = vec![];
        all.extend_from_slice(&v9_packet);
        all.extend_from_slice(&v5_packet);
        all.extend_from_slice(&v7_packet);
        all.extend_from_slice(&v9_packet);
        all.extend_from_slice(&ipfix_packet);
        all.extend_from_slice(&v5_packet);
        assert_yaml_snapshot!(NetflowParser::default().parse_bytes(&all));
    }

    #[test]
    fn it_parses_v9_many_flows() {
        let packet = [
            0, 9, 0, 3, 0, 0, 9, 9, 0, 1, 2, 3, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 16, 1, 2, 0,
            2, 0, 1, 0, 4, 0, 8, 0, 4, 1, 2, 0, 20, 9, 2, 3, 4, 9, 9, 9, 8, 9, 2, 3, 4, 9, 9,
            9, 8,
        ];
        assert_yaml_snapshot!(NetflowParser::default().parse_bytes(&packet));
    }

    #[test]
    fn it_parses_v9_options_template() {
        let packet = [
            0, 9, 0, 2, 0, 0, 9, 9, 0, 1, 2, 3, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 22, 1, 19, 0,
            4, 0, 8, 0, 2, 0, 2, 0, 34, 0, 2, 0, 36, 0, 1, 1, 19, 0, 9, 0, 2, 0, 100, 1,
        ];
        assert_yaml_snapshot!(NetflowParser::default().parse_bytes(&packet));
    }

    #[test]
    fn it_parses_v9_data_cached_template() {
        let packet = [
            0, 9, 0, 1, 0, 0, 9, 9, 0, 1, 2, 3, 0, 0, 0, 1, 0, 0, 0, 1, 1, 2, 0, 12, 9, 2, 3,
            4, 9, 9, 9, 8,
        ];
        let fields = vec![
            V9TemplateField {
                field_type_number: 1,
                field_type: crate::variable_versions::v9_lookup::V9Field::InBytes,
                field_length: 4,
            },
            V9TemplateField {
                field_type_number: 8,
                field_type: crate::variable_versions::v9_lookup::V9Field::Ipv4SrcAddr,
                field_length: 4,
            },
        ];
        let template = V9Template {
            field_count: 2,
            template_id: 258,
            fields,
        };
        let mut parser = NetflowParser::default();
        parser.v9_parser.templates.insert(258, template);
        assert_yaml_snapshot!(parser.parse_bytes(&packet));
    }

    #[test]
    fn it_parses_ipfix() {
        let packet = [
            0, 10, 0, 64, 1, 2, 3, 4, 0, 0, 0, 0, 1, 2, 3, 4, 0, 2, 0, 20, 1, 0, 0, 3, 0, 8, 0,
            4, 0, 12, 0, 4, 0, 2, 0, 4, 1, 0, 0, 28, 1, 2, 3, 4, 1, 2, 3, 3, 1, 2, 3, 2, 0, 2,
            0, 2, 0, 1, 2, 3, 4, 5, 6, 7,
        ];
        assert_yaml_snapshot!(NetflowParser::default().parse_bytes(&packet));
    }

    #[test]
    fn it_doesnt_panic_with_invalid_options_ipfix_template() {
        let packet = [
            0, 10, 0, 44, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 2, 0, 3, 0, 28, 1, 4, 0, 13, 0, 1,
            128, 123, 0, 4, 0, 0, 0, 2, 0, 41, 0, 2, 0, 42, 0, 2, 0, 0,
        ];
        NetflowParser::default().parse_bytes(&packet);
    }

    #[test]
    fn it_parses_ipfix_data_cached_template() {
        let packet = [
            0, 10, 0, 26, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 2, 0, 10, 0, 8, 0, 0, 1, 1,
        ];
        let fields = vec![
            IPFixTemplateField {
                field_type_number: 2,
                field_type:
                    crate::variable_versions::ipfix_lookup::IPFixField::PacketDeltaCount,
                field_length: 2,
                enterprise_number: None,
            },
            IPFixTemplateField {
                field_type_number: 8,
                field_type:
                    crate::variable_versions::ipfix_lookup::IPFixField::SourceIpv4address,
                field_length: 4,
                enterprise_number: None,
            },
        ];
        let template = IPFixTemplate {
            field_count: 2,
            template_id: 258,
            fields,
        };
        let mut parser = NetflowParser::default();
        parser.ipfix_parser.templates.insert(258, template);
        assert_yaml_snapshot!(parser.parse_bytes(&packet));
    }

    #[test]
    fn it_parses_ipfix_with_no_template_fields_raises_error() {
        let packet = [
            0, 10, 0, 26, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 2, 0, 10, 0, 8, 0, 0, 1, 1,
        ];
        let template = IPFixTemplate {
            field_count: 2,
            template_id: 258,
            fields: vec![],
        };
        let mut parser = NetflowParser::default();
        parser.ipfix_parser.templates.insert(258, template);
        assert_yaml_snapshot!(parser.parse_bytes(&packet));
    }

    #[test]
    fn it_parses_v9_with_no_template_fields_raises_error() {
        let packet = [
            0, 9, 0, 26, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 2, 0, 10, 0, 8, 0, 0, 1, 1,
        ];
        let template = V9Template {
            field_count: 2,
            template_id: 258,
            fields: vec![],
        };
        let mut parser = NetflowParser::default();
        parser.v9_parser.templates.insert(258, template);
        assert_yaml_snapshot!(parser.parse_bytes(&packet));
    }

    #[test]
    fn it_parses_ipfix_options_template() {
        let packet = [
            0, 10, 0, 44, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 2, 0, 3, 0, 28, 1, 4, 0, 3, 0, 1,
            128, 123, 0, 4, 0, 0, 0, 2, 0, 41, 0, 2, 0, 42, 0, 2, 0, 0,
        ];
        assert_yaml_snapshot!(NetflowParser::default().parse_bytes(&packet));
    }

    #[test]
    fn it_parses_ipfix_options_template_with_data() {
        let packet = [
            0, 10, 0, 64, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 2, 0, 3, 0, 28, 1, 4, 0, 3, 0, 1,
            128, 123, 0, 4, 0, 0, 0, 2, 0, 41, 0, 2, 0, 42, 0, 2, 0, 0, 1, 4, 0, 20, 0, 0, 0,
            1, 1, 20, 20, 20, 0, 0, 0, 2, 20, 20, 30, 30,
        ];
        assert_yaml_snapshot!(NetflowParser::default().parse_bytes(&packet));
    }
}
