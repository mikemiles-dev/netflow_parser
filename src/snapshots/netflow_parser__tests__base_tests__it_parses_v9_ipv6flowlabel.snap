---
source: src/tests.rs
expression: "NetflowParser::default().parse_bytes(&packets)"
---
- V9:
    header:
      version: 9
      count: 4
      sys_up_time: 10624535
      unix_secs: 1717095238
      sequence_number: 341
      source_id: 8
    flowsets:
      - header:
          flowset_id: 0
          length: 76
        body:
          templates:
            - template_id: 258
              field_count: 17
              fields:
                - field_type_number: 8
                  field_type: Ipv4SrcAddr
                  field_length: 4
                - field_type_number: 12
                  field_type: Ipv4DstAddr
                  field_length: 4
                - field_type_number: 15
                  field_type: Ipv4NextHop
                  field_length: 4
                - field_type_number: 7
                  field_type: L4SrcPort
                  field_length: 2
                - field_type_number: 11
                  field_type: L4DstPort
                  field_length: 2
                - field_type_number: 10
                  field_type: InputSnmp
                  field_length: 2
                - field_type_number: 14
                  field_type: OutputSnmp
                  field_length: 2
                - field_type_number: 252
                  field_type: Unknown
                  field_length: 4
                - field_type_number: 253
                  field_type: Unknown
                  field_length: 4
                - field_type_number: 2
                  field_type: InPkts
                  field_length: 4
                - field_type_number: 1
                  field_type: InBytes
                  field_length: 4
                - field_type_number: 22
                  field_type: FirstSwitched
                  field_length: 4
                - field_type_number: 21
                  field_type: LastSwitched
                  field_length: 4
                - field_type_number: 4
                  field_type: Protocol
                  field_length: 1
                - field_type_number: 5
                  field_type: SrcTos
                  field_length: 1
                - field_type_number: 256
                  field_type: Unknown
                  field_length: 2
                - field_type_number: 61
                  field_type: Direction
                  field_length: 1
      - header:
          flowset_id: 0
          length: 84
        body:
          templates:
            - template_id: 259
              field_count: 19
              fields:
                - field_type_number: 8
                  field_type: Ipv4SrcAddr
                  field_length: 4
                - field_type_number: 12
                  field_type: Ipv4DstAddr
                  field_length: 4
                - field_type_number: 15
                  field_type: Ipv4NextHop
                  field_length: 4
                - field_type_number: 7
                  field_type: L4SrcPort
                  field_length: 2
                - field_type_number: 11
                  field_type: L4DstPort
                  field_length: 2
                - field_type_number: 6
                  field_type: TcpFlags
                  field_length: 1
                - field_type_number: 10
                  field_type: InputSnmp
                  field_length: 2
                - field_type_number: 14
                  field_type: OutputSnmp
                  field_length: 2
                - field_type_number: 252
                  field_type: Unknown
                  field_length: 4
                - field_type_number: 253
                  field_type: Unknown
                  field_length: 4
                - field_type_number: 2
                  field_type: InPkts
                  field_length: 4
                - field_type_number: 1
                  field_type: InBytes
                  field_length: 4
                - field_type_number: 22
                  field_type: FirstSwitched
                  field_length: 4
                - field_type_number: 21
                  field_type: LastSwitched
                  field_length: 4
                - field_type_number: 4
                  field_type: Protocol
                  field_length: 1
                - field_type_number: 5
                  field_type: SrcTos
                  field_length: 1
                - field_type_number: 209
                  field_type: Unknown
                  field_length: 8
                - field_type_number: 256
                  field_type: Unknown
                  field_length: 2
                - field_type_number: 61
                  field_type: Direction
                  field_length: 1
      - header:
          flowset_id: 0
          length: 84
        body:
          templates:
            - template_id: 261
              field_count: 19
              fields:
                - field_type_number: 27
                  field_type: Ipv6SrcAddr
                  field_length: 16
                - field_type_number: 28
                  field_type: Ipv6DstAddr
                  field_length: 16
                - field_type_number: 62
                  field_type: Ipv6NextHop
                  field_length: 16
                - field_type_number: 7
                  field_type: L4SrcPort
                  field_length: 2
                - field_type_number: 11
                  field_type: L4DstPort
                  field_length: 2
                - field_type_number: 6
                  field_type: TcpFlags
                  field_length: 1
                - field_type_number: 10
                  field_type: InputSnmp
                  field_length: 2
                - field_type_number: 14
                  field_type: OutputSnmp
                  field_length: 2
                - field_type_number: 252
                  field_type: Unknown
                  field_length: 4
                - field_type_number: 253
                  field_type: Unknown
                  field_length: 4
                - field_type_number: 2
                  field_type: InPkts
                  field_length: 4
                - field_type_number: 1
                  field_type: InBytes
                  field_length: 4
                - field_type_number: 22
                  field_type: FirstSwitched
                  field_length: 4
                - field_type_number: 21
                  field_type: LastSwitched
                  field_length: 4
                - field_type_number: 4
                  field_type: Protocol
                  field_length: 1
                - field_type_number: 5
                  field_type: SrcTos
                  field_length: 1
                - field_type_number: 80
                  field_type: InDstMac
                  field_length: 6
                - field_type_number: 256
                  field_type: Unknown
                  field_length: 2
                - field_type_number: 61
                  field_type: Direction
                  field_length: 1
      - header:
          flowset_id: 0
          length: 88
        body:
          templates:
            - template_id: 262
              field_count: 20
              fields:
                - field_type_number: 27
                  field_type: Ipv6SrcAddr
                  field_length: 16
                - field_type_number: 28
                  field_type: Ipv6DstAddr
                  field_length: 16
                - field_type_number: 62
                  field_type: Ipv6NextHop
                  field_length: 16
                - field_type_number: 31
                  field_type: Ipv6FlowLabel
                  field_length: 3
                - field_type_number: 7
                  field_type: L4SrcPort
                  field_length: 2
                - field_type_number: 11
                  field_type: L4DstPort
                  field_length: 2
                - field_type_number: 6
                  field_type: TcpFlags
                  field_length: 1
                - field_type_number: 10
                  field_type: InputSnmp
                  field_length: 2
                - field_type_number: 14
                  field_type: OutputSnmp
                  field_length: 2
                - field_type_number: 252
                  field_type: Unknown
                  field_length: 4
                - field_type_number: 253
                  field_type: Unknown
                  field_length: 4
                - field_type_number: 2
                  field_type: InPkts
                  field_length: 4
                - field_type_number: 1
                  field_type: InBytes
                  field_length: 4
                - field_type_number: 22
                  field_type: FirstSwitched
                  field_length: 4
                - field_type_number: 21
                  field_type: LastSwitched
                  field_length: 4
                - field_type_number: 4
                  field_type: Protocol
                  field_length: 1
                - field_type_number: 5
                  field_type: SrcTos
                  field_length: 1
                - field_type_number: 80
                  field_type: InDstMac
                  field_length: 6
                - field_type_number: 256
                  field_type: Unknown
                  field_length: 2
                - field_type_number: 61
                  field_type: Direction
                  field_length: 1
- V9:
    header:
      version: 9
      count: 2
      sys_up_time: 10724622
      unix_secs: 1717095339
      sequence_number: 356
      source_id: 8
    flowsets:
      - header:
          flowset_id: 258
          length: 102
        body:
          data:
            data_fields:
              - 0:
                  - Ipv4SrcAddr
                  - Ip4Addr: 192.168.18.10
                1:
                  - Ipv4DstAddr
                  - Ip4Addr: 141.24.12.2
                2:
                  - Ipv4NextHop
                  - Ip4Addr: 0.0.0.0
                3:
                  - L4SrcPort
                  - DataNumber: 49856
                4:
                  - L4DstPort
                  - DataNumber: 53
                5:
                  - InputSnmp
                  - DataNumber: 2
                6:
                  - OutputSnmp
                  - DataNumber: 0
                7:
                  - Unknown
                  - Vec:
                      - 0
                      - 0
                      - 0
                      - 0
                8:
                  - Unknown
                  - Vec:
                      - 0
                      - 0
                      - 0
                      - 0
                9:
                  - InPkts
                  - DataNumber: 1
                10:
                  - InBytes
                  - DataNumber: 82
                11:
                  - FirstSwitched
                  - Duration:
                      secs: 10690
                      nanos: 167000000
                12:
                  - LastSwitched
                  - Duration:
                      secs: 10690
                      nanos: 167000000
                13:
                  - Protocol
                  - ProtocolType: Udp
                14:
                  - SrcTos
                  - DataNumber: 0
                15:
                  - Unknown
                  - Vec:
                      - 8
                      - 0
                16:
                  - Direction
                  - DataNumber: 1
              - 0:
                  - Ipv4SrcAddr
                  - Ip4Addr: 192.168.18.10
                1:
                  - Ipv4DstAddr
                  - Ip4Addr: 141.24.12.2
                2:
                  - Ipv4NextHop
                  - Ip4Addr: 0.0.0.0
                3:
                  - L4SrcPort
                  - DataNumber: 40483
                4:
                  - L4DstPort
                  - DataNumber: 53
                5:
                  - InputSnmp
                  - DataNumber: 2
                6:
                  - OutputSnmp
                  - DataNumber: 0
                7:
                  - Unknown
                  - Vec:
                      - 0
                      - 0
                      - 0
                      - 0
                8:
                  - Unknown
                  - Vec:
                      - 0
                      - 0
                      - 0
                      - 0
                9:
                  - InPkts
                  - DataNumber: 1
                10:
                  - InBytes
                  - DataNumber: 82
                11:
                  - FirstSwitched
                  - Duration:
                      secs: 10688
                      nanos: 891000000
                12:
                  - LastSwitched
                  - Duration:
                      secs: 10688
                      nanos: 891000000
                13:
                  - Protocol
                  - ProtocolType: Udp
                14:
                  - SrcTos
                  - DataNumber: 0
                15:
                  - Unknown
                  - Vec:
                      - 8
                      - 0
                16:
                  - Direction
                  - DataNumber: 1
      - header:
          flowset_id: 262
          length: 99
        body:
          data:
            data_fields:
              - 0:
                  - Ipv6SrcAddr
                  - Ip6Addr: "fd01:8::2a20:235f:1f7b:9379"
                1:
                  - Ipv6DstAddr
                  - Ip6Addr: "fd00::b2f2:8ff:fe20:1180"
                2:
                  - Ipv6NextHop
                  - Ip6Addr: "::"
                3:
                  - Ipv6FlowLabel
                  - DataNumber: 128630
                4:
                  - L4SrcPort
                  - DataNumber: 58037
                5:
                  - L4DstPort
                  - DataNumber: 53
                6:
                  - TcpFlags
                  - DataNumber: 0
                7:
                  - InputSnmp
                  - DataNumber: 2
                8:
                  - OutputSnmp
                  - DataNumber: 0
                9:
                  - Unknown
                  - Vec:
                      - 0
                      - 0
                      - 0
                      - 0
                10:
                  - Unknown
                  - Vec:
                      - 0
                      - 0
                      - 0
                      - 0
                11:
                  - InPkts
                  - DataNumber: 1
                12:
                  - InBytes
                  - DataNumber: 102
                13:
                  - FirstSwitched
                  - Duration:
                      secs: 10688
                      nanos: 891000000
                14:
                  - LastSwitched
                  - Duration:
                      secs: 10688
                      nanos: 891000000
                15:
                  - Protocol
                  - ProtocolType: Udp
                16:
                  - SrcTos
                  - DataNumber: 0
                17:
                  - InDstMac
                  - MacAddr: "10:90:27:E0:43:6D"
                18:
                  - Unknown
                  - Vec:
                      - 134
                      - 221
                19:
                  - Direction
                  - DataNumber: 1
